### ISSUE TITLE
Predict bbox in training mode

### ISSUE THREAD

**Issue by TWDH** [xingyizhou/CenterNet#730]
@xingyizhou Hi dalao, I was trying to use centernet as RPN to generate rois and feed into RCNN header and did much experiment. I found that when using the "model.train()" mode to predict the bbox is very different from when I predict the same picture using demo.py[eval.()], Finally I figured out that ".train()" use batchNormalization and dropout while when testing or demo, the model turns to ".eval()" which doesn't use batchNormalization and dropout, and this make the bbox result different from each other.
However, considering using centernet as RPN has to be in ".train()" mode, is there a way to get the same result as in ".eval()" model? I tried increase batch size=8 that kind of make the gap smaller, but still has the problem, Hope you can comment on it, Thanks very much!

**Comment by xingyizhou** [comment]
Thank you for your experiments. We are trying using CenterNet as RPN too and had some positive results. It shouldn't be too hard to make it work. Unfortunately we don't plan to make the exact method public now. What I can tell is it's completely fine to use model.train() for generating proposals, as the default RPN also operates this way.

**Comment by TWDH** [comment]
> Thank you for your experiments. We are trying using CenterNet as RPN too and had some positive results. It shouldn't be too hard to make it work. Unfortunately we don't plan to make the exact method public now. What I can tell is it's completely fine to use model.train() for generating proposals, as the default RPN also operates this way.

Hi @xingyizhou, thanks for replying. I'm not sure if you would like to share some training details cause I have some questions that confused me a lot during the experiment, appriciate it if you can give me some hints :)

I used DlaNet as backbone and following the procedure in "detector.run()" to predict the bbox and feed it into ROI-Align then go to 2 FC layers and separate into RCNN_cls and RCNN_bbox, and training use [loss = hm + wh + offset + rcnn_cls + rcnn_bbox] with [learning rate= 1.25e-4] (one RTX 2080super + pytorch1.1 + CUDA10), I trained on the pretrained network that you provided for 25 epochs. Then, I repredict the whole thing based on rcnn(cls+bbox) output and unfortunately it not only does harms to the original CenterNet about 4~5% decrese but also the re-predict result is bad. 

Q1:  I think there might be a problem of learning rate as the pretrained dlanet already decrase to 1.25e-6, or it could be the mis-design of loss function which decrease the whole thing? It's weird that the training loss is going down every epoch while the validation loss continuously going up. Do you think I should train from scratch or the whole strategy is wrong?

Q2: I also tried training with [loss = rcnn_cls + rcnn_bbox] and fix the CenterNet parameters. This time, I use the predicted 100 bbox from CenterNet and just repredict  bbox locations. it gives me pretty much the same mAP as the original one with 0.1~0.2% decrease, may be only refine the bbox location doesn't help much?

Generally speaking, using RCNN to repredict result should give some inprovment but as the Centernet's prediction has a lot of false positive or not detected objects, use the RCNN head can only select objects in the given rois(centernet generated) which won't give a big difference or we should use the loss(maybe mask loss, or cls loss) to refine the original CenterNet directly which make more sense, just like CenterMask recently published?
Sorrrrrrrrrrry to trouble you too much, but I'm very desperate to know if I did anything wrong, THANKS dalao! ~~~

**Comment by xingyizhou** [comment]
Thank you for your detailed discussion. Unfortunately, I really should not disclose our ongoing project or look at your ideas (I haven't 🙈 ), especially publicly. Here are all the hints I can give: we used all the FasterRCNN settings, just replace RPN with CenterNet in detectron2 (you need to implement CenterNet in detectron2 for doing this).

**Comment by TWDH** [comment]
> Thank you for your detailed discussion. Unfortunately, I really should not disclose our ongoing project or look at your ideas (I haven't 🙈 ), especially publicly. Here are all the hints I can give: we used all the FasterRCNN settings, just replace RPN with CenterNet in detectron2 (you need to implement CenterNet in detectron2 for doing this).

@xingyizhou That's very kind of you, many Thanks

**Comment by pplkjh** [comment]
That was the good point @xingyizhou. you saved my whole projects!. I had a similar concept of the problem in my model to test. using 1 batch to train to see how comes the loss looks like, and put the train-set image into the model with an overtrained weight for the test printing the loss either. compared both losses, then found that it has a large gap between them in the loss of the same image to train and eval. I looked up your issue and I set off the model.eval() and it looks much like the prediction I expected from the training. 
I would try re-train my model with small-batch and test with no model.eval().
I hope you have a good result.

**Comment by TWDH** [comment]
@pplkjh Hi, did you use CenterNet as your code or detectron2 to rebuild it? it don't see much difference between them but my training result still doesn't show any improvement. I just trained another 20epoch with "rcnn_cls_loss" based on the pretrained model of dlanet(140 epoch), would you like to share your result? thanks

**Comment by pplkjh** [comment]
morning, @xingyizhou . I tried some version of my model using mobile-net v3 or YOLO4 as the backbone. so I trained small datasets from the beginning without any pre-trained weight. 
why don't you try a test with the same batch of the input tensor? 
Ex) train input tensor size : [8, 3, 420, 420] as [batch size, RGB, height, width]
then try a test with the same input tensor size like [8, 3, 420, 420], with no model.eval(). 
It at least worked in my problem.

**Comment by TWDH** [comment]
@pplkjh That seems to be a good result, so you didn’t train it on coco? Actually, what’s your batch size and learning rate? I realize it’s probably because the wrong lr lead to the model not converge.  In addition, did you take any code as reference in terms of fasterRcnn part(roi align and rcnn), cause I follow jwyang’s roi and rcnn structure to build my own model. Glad if you can share
